### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by sphamba**
# Prework

- [x] I understand and agree to this repository's [code of conduct](https://github.com/ghiggi/ximage/blob/main/CODE_OF_CONDUCT.md).
- [x] I understand and agree to this repository's [contributing guidelines](https://github.com/ghiggi/ximage/blob/main/CONTRIBUTING.md).
- [ ] I have already submitted an [issue](https://github.com/ghiggi/ximage/issues) or [discussion thread](https://github.com/ghiggi/ximage/discussions) to discuss my idea with the maintainers.

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Documentation
- [ ] Tutorial
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [x] Other, please describe: Tests

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and communicate accordingly:

**The PR fulfills these requirements:**

- [x] It's submitted to the branch named as follow :
  - Fix a bug: `bugfix-<some_key>-<word>`
  - Improve the doc: `doc-<some_key>-<word>`
  - Improve a tutorial `tutorial-<some_key>-<word>`
  - Add a new feature: `feature-<some_key>-<word>`
  - Refactor some code: `refactor-<some_key>-<word>`
  - Optimize some code: `optimize-<some_key>-<word>`
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] Don't forget to link PR to issue if you are solving one.
- [x] All tests are passing.
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

# Related GitHub issues and pull requests

- Ref: #

# Summary

Added tests for:
- labels/labels.py
- patch/checks.py
- utils/checks.py

### CODE REVIEW

File: ximage/utils/checks.py [hunk 1]
```diff
@@ -27,6 +27,10 @@ def are_all_integers(arr, negative_allowed=True, zero_allowed=True):
         True if all values in the array are integers, False otherwise.
 
     """
+    # Convert to numpy array
+    if not isinstance(arr, np.ndarray):
```
Comment (ghiggi) [inline ximage/utils/checks.py:31]: `np.asanyarray` should do the job here.
`np.array` always copy the data !

File: ximage/tests/labels/test_labels.py [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+    dataset = xr.Dataset({label_name: labels_array})
+    dataset_returned = labels.highlight_label(dataset, label_name, label_id)
+    labels_array_expected = np.array(
+        [
+            [0, 0, 2],
+            [0, 0, 2],
+            [0, 0, 0],
+        ]
+    )
+    print(dataset_returned)
```
Comment (evanjt) [inline ximage/tests/labels/test_labels.py:169]: Could be more better placed within the assertion output on fail
```
assert np.array_equal(dataset_returned[label_name], labels_array_expected), (
    f"Returned {dataset_returned} but expected {labels_array_expected}"
)
```

File: ximage/tests/labels/test_labels.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+
+    # Check exception raised if label_indices is empty
+    with pytest.raises(ValueError):
+        labels._np_redefine_label_array(array, np.array([]))
+
+
+def test_xr_redefine_label_array():
+    """Test _xr_redefine_label_array"""
+
+    array = xr.DataArray(
```
Comment (evanjt) [inline ximage/tests/labels/test_labels.py:469]: This array is being commonly used, I would suggest defining this as a fixture in `conftest.py`, and even better labelling it something nice if it represents a common array to be used.

```
@pytest.fixture
def common_3_by_3_matrix() -> xr.DataArray:
        return  xr.DataArray(
        [
            [0, 1, 2],
            [2, 3, 3],
            [3, 0, 0],
        ]
    )
```

then:
```
def test_xr_redefine_label_array(common_3_by_3_matrix: xr.DataArray):
    ... 
```

**Review by evanjt — APPROVED** [review]

File: ximage/tests/labels/test_labels.py (hunk 2 above)
Comment (sphamba) [reply ximage/tests/labels/test_labels.py:169]: Indeed, I forgot to remove the print statement

File: ximage/tests/labels/test_labels.py (hunk 3 above)
Comment (sphamba) [reply ximage/tests/labels/test_labels.py:469]: I thought about it, but then I found myself creating several different matrices for testing different features. Also, I find it nice to directly see what the matrix looks like in the test. What do you think?

Comment (ghiggi) [reply ximage/tests/labels/test_labels.py:468]: For me fine to just define close to the test. 
Maybe you can define a fixture for an array representing edge cases: i.e all zeros, all.nan, all 1 ... 
I don't have a strong feeling about that.

BTW: If you use fixtures, would be good if at the top of the file we list the fixture that we use ;)

Comment (evanjt) [reply ximage/tests/labels/test_labels.py:468]: Both approaches are for me. If the array is defining a feature of something specific that is common to all tests, then it would be useful for it to be a fixture.

**Review by evanjt — APPROVED** [review]

**Comment by sphamba** [comment]
Added `tests/patch/test_slices.py`

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -3,41 +3,47 @@ on: [push, pull_request, workflow_dispatch]
 
 jobs:
   tests:
-    name: ${{ matrix.os }}, Python ${{ matrix.python-version }}${{ matrix.experimental && ', experimental' || '' }}
+    name: ${{ matrix.os }}, Python ${{ matrix.python-version }}${{ matrix.experimental && ', latest deps' || '' }}
```
Comment (ghiggi) [inline .github/workflows/tests.yml:6]:
```suggestion
    name: ${{ matrix.os }}, Python ${{ matrix.python-version }}${{ matrix.experimental && ', latest_dependencies' || '' }}

```

File: ximage/labels/labels.py [hunk 5]
```diff
@@ -129,6 +130,7 @@ def _get_label_value_stats(arr, label_arr, label_indices=None, stats="area"):
             index=label_indices,
             func=stats,
             out_dtype=float,
+            default=None,
```
Comment (ghiggi) [inline ximage/labels/labels.py:133]: Maybe we could regroup `out_dtype`, `default `and `pass_positions` inside a `labeled_comprehensions_kwargs `dictionary ... here use `**labeled_comprehensions_kwargs` (which if `{}` takes the current defaults) and add the argument `labeled_comprehensions_kwargs={}` to `_get_labels_stats`, `_get_labels`, `_xr_get_labels`, `label` (and the accessor)

File: ximage/patch/slices.py [hunk 6]
```diff
@@ -186,9 +190,13 @@ def enlarge_slices(list_slices, min_size, valid_shape):
         min_size = [min_size] * len(list_slices)
     if isinstance(valid_shape, int):
         valid_shape = [valid_shape] * len(list_slices)
-    if isinstance(min_size, (list, tuple)) and len(min_size) != len(min_size):
+    if isinstance(min_size, (list, tuple)) and len(min_size) != len(list_slices):
```
Comment (ghiggi) [inline ximage/patch/slices.py:193]: Aahaha ops ... `len(min_size) != len(min_size)` was kinda funny 😄

File: ximage/patch/slices.py [hunk 7]
```diff
@@ -225,20 +233,22 @@ def get_slice_around_index(index, size, min_start=0, max_stop=np.inf):
     Returns
     -------
     slice
-        The new slice object with a size of at least min_size and respecting the left and right bounds.
+        A slice object with the desired size and respecting the left and right bounds.
```
Comment (ghiggi) [inline ximage/patch/slices.py:237]: Should we add that the original slice is returned if the slice is already larger than `size `?

File: ximage/patch/slices.py [hunk 8]
```diff
@@ -375,7 +386,7 @@ def get_partitions_slices(
     idxs = _get_partitioning_idxs(
         start=start, stop=stop, stride=stride, slice_size=slice_size, method=method
     )
-    slices = [slice(idxs[i], idxs[i + 1], slice_step) for i in range(len(idxs) - 1)]
+    slices = [slice(idxs[i], idxs[i] + slice_size, slice_step) for i in range(len(idxs) - 1)]
```
Comment (ghiggi) [inline ximage/patch/slices.py:389]: Excellent spot !

File: ximage/tests/labels/test_labels.py [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+        ]
+    )
+
+    values_area = [5, 1, 3]
+
+    # With default labels indices (all) and stats (area)
+    values = labels._get_label_value_stats(array, label_array)
+    assert np.array_equal(values, values_area)
+
+    # custob stats function
```
Comment (ghiggi) [inline ximage/tests/labels/test_labels.py:298]:
```suggestion
    # custom stats function
```

File: ximage/tests/labels/test_labels.py [hunk 10]
```diff
@@ -0,0 +1,735 @@
...
+        array, label_array, sort_decreasing=False
+    )
+    assert np.array_equal(label_indices, np.array([1, 2, 0]))
+    assert np.array_equal(values_returned, [1, 3, 5])
+
+
+def test_vec_translate():
+    """Test _vec_translate"""
+
+    array = np.array(
```
[the remark this answers is not in the document]
Comment (sphamba) [reply ximage/tests/labels/test_labels.py:339]: I added a check that an error is raised in these cases. For now, even if a single key is not present in the array, it raises an error. The behavior could be changed to just remove it and keep the other valid keys, but this is not a situation that should arise I think

File: ximage/labels/labels.py (hunk 5 above)
Comment (sphamba) [reply ximage/labels/labels.py]: Done

**Comment by ghiggi** [comment]
@sphamba yes you are right ! Right now we increment the counter by 1 while:
- for a given label multiple patches can be extracted when tiling or sliding
- we could also have the case where no patch is extracted for a label (maybe with a custom center on method?)

Let's discuss tomorrow how to approach these cases :eye:

**Review by ghiggi — APPROVED** [review]
Just perfect !

File: ximage/tests/labels/test_labels.py [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+
+    # Check conversion to np.ndarray
+    array = np.ones((3, 3))
+
+    def check(array):
+        checked_array = labels._check_array(array)
+        assert isinstance(checked_array, np.ndarray)
+        assert np.array_equal(checked_array, array)
+
+    apply_to_all_array_types(check, array)
```
Comment (ghiggi) [inline ximage/tests/labels/test_labels.py:229]: There is a small fix to be done here. 
`C:\Users\runneradmin\micromamba\envs\ximage\Lib\site-packages\dask\array\core.py:1711: FutureWarning: The `numpy.array_equal` function is not implemented by Dask array. You may want to use the da.map_blocks function or something similar to silence this warning. Your code may stop working in a future release.`

File: ximage/tests/patch/test_labels_patch.py [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+    """Test _get_point_with_max_value"""
+
+    shape = (4, 5)
+    array = np.random.rand(*shape)
+    max_value = np.max(array)
+
+    returned_point = labels_patch._get_point_with_max_value(array)
+    assert array[returned_point] == max_value
+
+    # Check None return if only nan in array
```
Comment (ghiggi) [inline ximage/tests/patch/test_labels_patch.py:585]: Here capture the `RuntimeWarning: All-NaN slice encountered point = np.argwhere(arr == np.nanmax(arr))` so to have clean the warning summary of the CI ;)

File: ximage/tests/patch/test_labels_patch.py [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+
+    shape = (4, 5)
+    array = np.random.rand(*shape)
+    min_value = np.min(array)
+
+    returned_point = labels_patch._get_point_with_min_value(array)
+    assert array[returned_point] == min_value
+
+    # Check None return if only nan in array
+    array = np.full(shape, np.nan)
```
Comment (ghiggi) [inline ximage/tests/patch/test_labels_patch.py:601]: Here capture the RuntimeWarning: All-NaN slice encountered point = np.argwhere(arr == np.nanmax(arr)) so to have a clean the warning summary of the CI.
Or we can capture the warning already in the code, since we allow this behaviour ;)

File: ximage/tests/accessor/test_accessor_args.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+    modified_default_arguments = {**base_default_arguments, **changed_default_arguments}
+
+    assert modified_default_arguments == reference_default_arguments
+
+
+# Tests for accessors' methods default arguments ###############################
+
+
+def test_base_accessor_label():
```
Comment (ghiggi) [inline ximage/tests/accessor/test_accessor_args.py:40]: Here we could do something more elaborate that we can also use in the GPM-API software.
For each class accessor, we retrieve the list of methods, loop over it, and call  `compare_default_arguments`.
So that if new methods are added, are automatically tested ;)

Comment (sphamba) [reply ximage/tests/accessor/test_accessor_args.py:40]: @ghiggi Yes I can attempt something. I could use `inspect.getsourcelines` to retrieve the path and name of the imported method, which could be a bit dirty. Would you have another way of doing this in mind?

Comment (ghiggi) [reply ximage/tests/accessor/test_accessor_args.py:40]: No. I was thinking to the same approach you detailed. 
And also let's do a test where we list the currently implemented methods, and we check that they exists in the accessor, so that we can better track future changes in methods' names.

Comment (sphamba) [reply ximage/tests/accessor/test_accessor_args.py:40]: The thing is that for now, I am not aware of a way to find out that, for example, the `label_patches` accessor's method must call `get_patches_from_labels` (a different method name) from `ximage.patch.labels_patch` (an arbitrary path), without needing to read the source code of the accessor's method.

Also, would you see an issue in rewriting the accessor's method definition as something along the lines of
```
from ximage.patch.labels_patch import get_patches_from_labels
self.label_patches = functools.partial(get_patches_from_labels, self._obj)
```
? I am not 100% sure it works, but it would remove the need from having to copy the list of arguments.

Comment (ghiggi) [reply ximage/tests/accessor/test_accessor_args.py:40]: Mmh we can try out. 
I propose the following @sphamba:
- we merge the current PR 
- we address the testing of the accessors in a separate PR (where I can also experiment a couple of stuff ...)

Do you agree?
We could explore doing the method assignment in the `__init__` maybe, and also copy the docstring to the method!

Comment (sphamba) [reply ximage/tests/accessor/test_accessor_args.py:40]: Sounds good! Ready to merge
